packages/core/react/src/WalletProvider.tsx

```diff
@@ -23,285 +24,158 @@ export interface WalletProviderProps {
-                    : {
-                          adapter: adapter,
-                          readyState: adapter.readyState,
-                      };
-            })
-        );
+function getIsMobile(adapters: Adapter[]) {
+    const userAgentString = getUserAgent();
+    return getEnvironment({ adapters, userAgentString }) === Environment.MOBILE_WEB;
+}
 
-        function handleReadyStateChange(this: Adapter, readyState: WalletReadyState) {
-            setWallets((prevWallets) => {
-                const index = prevWallets.findIndex(({ adapter }) => adapter === this);
-                if (index === -1) return prevWallets;
+function getUriForAppIdentity() {
+    const location = globalThis.location;
+    if (location == null) {
+        return;
+    }
+    return `${location.protocol}//${location.host}`;
+}
 
-                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-                const { adapter } = prevWallets[index]!;
-                return [...prevWallets.slice(0, index), { adapter, readyState }, ...prevWallets.slice(index + 1)];
-            });
+export function WalletProvider({
+    autoConnect,
+    localStorageKey = 'walletName',
+    wallets: adapters,
+    ...props
+}: WalletProviderProps) {
+    const { connection } = useConnection();
+    const mobileWalletAdapter = useMemo(() => {
+        if (!getIsMobile(adapters)) {
+            return null;
         }
-
-        adapters.forEach((adapter) => adapter.on('readyStateChange', handleReadyStateChange, adapter));
-        return () => adapters.forEach((adapter) => adapter.off('readyStateChange', handleReadyStateChange, adapter));
-    }, [adapters]);
-
-    // When the selected wallet changes, initialize the state
-    useEffect(() => {
-        const wallet = name && wallets.find(({ adapter }) => adapter.name === name);
-        if (wallet) {
-            setState({
-                wallet,
-                adapter: wallet.adapter,
-                connected: wallet.adapter.connected,
-                publicKey: wallet.adapter.publicKey,
-            });
-        } else {
-            setState(initialState);
+        const existingMobileWalletAdapter = adapters.find(
+            (adapter) => adapter.name === SolanaMobileWalletAdapterWalletName
+        );
+        if (existingMobileWalletAdapter) {
+            return existingMobileWalletAdapter;
         }
-    }, [name, wallets]);
-
-    // If the window is closing or reloading, ignore disconnect and error events from the adapter
-    useEffect(() => {
-        function listener() {
-            isUnloading.current = true;
+        return new SolanaMobileWalletAdapter({
+            addressSelector: createDefaultAddressSelector(),
+            appIdentity: {
+                uri: getUriForAppIdentity(),
+            },
+            authorizationResultCache: createDefaultAuthorizationResultCache(),
+            cluster: getInferredClusterFromEndpoint(connection?.rpcEndpoint),
+        });
```

Comment:
This all makes sense, though I do wonder if maybe we want to provide an easy way for dapps to customize their MWA without instantiating it (name and icon, mostly) and the maybe add a `cluster` prop to `ConnectionProvider` which wouldn't hurt in general to have here, or for the Wallet Standard.

Reply:
I'm hesitant to add things (at all, but particularly) to `ConnectionProvider` that only work for one platform. Like, adding `icon` doesn't _do_ anything on desktop, which might be confusing for folks.

Reminds me though, I wanted to suggest that we conventionally look for an icon at the root of the site if one isn't supplied; favicon style.

Reply:
Ah, what I meant is, adding optional cluster to ConnectionProvider would actually be _extremely_ useful for the Wallet Standard as well (which also infers it, messily). So it would do something useful on mobile and desktop.

Reply:
Interesting. Well, let's land both and then we can figure out how to patch them both up at once, if at all. I think I'm pro:

```typescript
type ClusterfulEndpoint = Readonly<{
  cluster: Cluster,
  url: string,
}>;

type ConnectionProviderProps = Readonly<{
  endpoint: string | ClusterfulEndpoint,
  /* ... */
}>;
```

Something like that.

Reply:
lol @ `Clusterful`